### ISSUE TITLE
Linux Help?

### ISSUE THREAD

**Issue by PixelTheGreat** [GrowtopiaNoobs/GrowtopiaServer#57]
So i used wine, i downloaded your server.exe cause i can't or don't know how to compile the sln or the `enet server test.cpp` file on Linux, anyways do i need to use run.py or just directly use wine server.exe ?

**Comment by willi123yao** [comment]
Windows executables can't be ran on Linux without any emulation(VMs) or compatibility layer such as wine,  never tried running it under wine, you can attempt it but you will also require the vc_redist installation to support running programs built with Visual Studio C++ 2015. 

It is possible rebuild the server for linux, however you will need to make some changes to the source code before compiling, notably such as removing windows.h as one of the header files (this will remove autosave and world save feature sadly), as well as replacing the encryption functions used for storing user passwords since I couldn't get it to compile as it was missing the library, and finally some slight changes to the code to fix some errors (the linux compiler is slightly more strict than Visual Studio). For the actual compilation, normal GNU linux build tools will do, such as gcc and g++.

I might make a fork of this project and support compiling for Linux/MacOS though, depending if I have the time to maintain it.

**Comment by willi123yao** [comment]
Also just to add on, the run.py starts the webserver which is required for Growtopia to obtain the server address (The "Getting server address..." part that you see every time you login to Growtopia), otherwise if you only changed your hosts file it will cause Growtopia to crash when logging in.

**Comment by PixelTheGreat** [comment]
How about the Current Linux Rewrite for the Server? Will that work?

**Comment by willi123yao** [comment]
Which linux rewrite are you referring to?

**Comment by GrowtopiaNoobs** [comment]
I know currently only about this attept to rewrite it for linux: https://github.com/coldchip/growtopiaserver but it is lacking alot of featueres. Btw good luck with rewriting server for linux :+1: Kinda funny is that bcrypt library was orginally for linux only, so I needed to rewrite it for windows. (Here is link to original library: https://github.com/rg3/libbcrypt ) It is using /dev/urandom and some POSIX functions.

**Comment by willi123yao** [comment]
Oh thanks for providing me with the library for hashing, I have been successful in making a linux/MacOS version of the server. I used the linux version of bcrypt to enable password hashing, as well as fixed some bugs that you have left out (will send in a PR soon), and also reimplemented world saving upon exit so now pretty much it has all the features from your code, just that it is also able to compile on linux/MacOS. I am still looking to see if it can be compiled on Windows, most likely it should be possible without any problems, and I'll update you guys again later on if I do.

Link to my linux-compatible version: https://github.com/willi123yao/GrowtopiaServer

**Comment by willi123yao** [comment]
UPDATE: I have managed to compile my edit on windows successfully, however I have decided to ditch bcrypt entirely since there are many compilation issues, and to use SHA256 instead. I have also provided precompiled executables for the various platforms over on my [Releases page](https://github.com/willi123yao/GrowtopiaServer/releases). 

The windows version also has no dependencies for Visual Studio, which means it is able to run without installing vc_redist (for anyone who has an older operating system)

**Comment by willi123yao** [comment]
@GrowtopiaNoobs Not sure if you want to check it out, only changed a few things, the rest are the same as your code.

**Comment by GrowtopiaNoobs** [comment]
I looked at it and I would just say, that I used bcrypt, beucase it is hard to crack those hashes. So I would recomend you to use salt and maybe multiple itterations of hash function when you are using fast hash functions.

**Comment by willi123yao** [comment]
Yeah I agree, for now its just simple hashing to make sure it works across platforms, and it also uses constant time comparison for extra security.

The hash cannot be accessed outside of actual server, unless you get hacked into or have a data breach.

**Comment by MarmoriPelaa** [comment]
I made it work on my raspberry pi running Linux based OS. Mostly, the code there is compatible, though there are a few functions that are only on windows (which I ported for Linux). There is lot of other stuff added in the code (such as a randomly running gem event, lol), so I don't know if you want the messy code that I am working on. Here are some tips:
- Use C++11
- Remove libraries as necessary. (I removed the password hashing completely for easy modification of broken passwords. Also the hashing function used bcrypt which I was unable to get on Linux.)
- Google a lot, and use sites like [stackoverflow](https://stackoverflow.com) as your help.
Wrapping it up, if you know what you're doing, you should be able to get it up and running in less than 3 days.

**Comment by GrowtopiaNoobs** [comment]
@MarmoriPelaa For sure send us any code you have. You can put it publicly on github or change it a bit and make PR with all features (maybe not that raw password feature). And that bcrypt library is originally for linux, but I ported it to windows, so if you will google a bit you would for sure found original one. Good job :+1:

**Comment by OverallGod** [comment]
@GrowtopiaNoobs All I really did was change the encryption to sha256 and a couple more things. And its running perfectly on linux

**Comment by OverallGod** [comment]
Actually if you look at the screenshots on my most recent PR you will see that I'm using Linux mint
